tests/shared/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala

```diff
@@ -1933,4 +1933,161 @@ class SyntacticSuite extends scala.meta.tests.parsers.ParseSuite {
     assertSyntax(tree, "\"\"\"\n\\\"\\\"\\\"\"\"\"")(tree)
   }
 
+  test("#2046 new line separator required before some Term.Block") {
+    assertEquals(
+      blockStat("{val a = 1; {b}; val c = 1}").syntax,
+      s"""
+      |{
+      |  val a = 1
+      |
+      |  {
+      |    b
+      |  }
+      |  val c = 1
+      |}
+      """.trim.stripMargin.replace("\n", EOL)
+    )
+    assertEquals(
+      blockStat("{val a = 1; {b}; {c}}").syntax,
+      s"""
+      |{
+      |  val a = 1
+      |
+      |  {
+      |    b
+      |  }
+      |  {
+      |    c
+      |  }
+      |}
+      """.trim.stripMargin.replace("\n", EOL)
+    )
+    assertEquals(
+      blockStat("{{a}; val b = 1; val c = 2; {d}}").syntax,
+      s"""
+      |{
+      |  {
+      |    a
+      |  }
+      |  val b = 1
+      |  val c = 2
+      |
+      |  {
+      |    d
+      |  }
+      |}
+      """.trim.stripMargin.replace("\n", EOL)
+    )
+
+    def testBlock(t: Stat, expectedStr: String, needSep: Boolean): Unit = {
+      def expected(sep: String): String =
+        s"""
+        |{
+        |${expectedStr.split("\n").map("  " + _).mkString("", "\n", sep)}
+        |  {
+        |    a
+        |  }
+        |}
+        """.trim.stripMargin.replace("\n", EOL)
+      val newBlock = q"$t;{a}"
+      assertEquals(
+        newBlock.syntax,
+        expected(sep = if (needSep) "\n" else "")
+      )
+      assertEquals(
+        newBlock.structure,
+        templStat(newBlock.syntax).structure
+      )
+      // If it compiles it cannot produce the same tree, otherwise the separator has unnecessary
+      if (needSep)
+        scala.util.Try(templStat(expected(sep = ""))).toOption.foreach { statNoSep =>
+          assertNotEquals(
+            newBlock.structure,
+            statNoSep.structure
+          )
+        }
+    }
+
+    def sep(t: Stat, expected: String): Unit = testBlock(t, expected, needSep = true)
+    def noSep(t: Stat, expected: String): Unit = testBlock(t, expected, needSep = false)
+
+    def declTest(f: String => Unit): Unit = Seq("val", "var", "def").foreach(f)
+    def classTest(f: String => Unit): Unit = Seq("class", "object", "trait").foreach(f)
+
+    declTest(k => sep(templStat(s"$k foo: Int"), s"$k foo: Int"))
```

Comment:
for every case where `guessXxx` returns true, could we add a test, please (unless it's already been done)?

Reply:
The current tests should cover every case. The cases that are constant (true or false) there is only one case, for the remaining cases I created a positive and a negative test.